### PR TITLE
Currency: Add interactive UI

### PR DESCRIPTION
**Pull request by pjhampton**
## Description of new Instant Answer, or changes

Adds interactive UI to the Currency IA

### SS1. Normal theme
screen shot 2017-05-18 at 5 54 57 pm

### SS2. Dark Theme
screen shot 2017-05-18 at 5 54 18 pm

### SS3. Mobile / Dark Theme
screen shot 2017-05-18 at 5 54 35 pm

## Related Issues and Discussions

Internal Discussions

## People to notify

@moollaza 
@bsstoner 

---

Instant Answer Page: https://duck.co/ia/view/currency

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by pjhampton** [comment]
Anybody interested in reviewing? @duckduckgo/duckduckhack-contributors 
You can test this PR by running:

```
$ g fetch origin pull/3294/head:pjhampton/pjhampton/Currency_add-interactive-ui
$ g checkout pjhampton/Currency_add-interactive-ui
$ duckpan server currency
```

**Comment by TheOdd** [comment]
@pjhampton I know I can't officially review it, but I did pull the request down and looked it over. Looks good from what I can tell!

File: share/spice/currency/currency.js [hunk 1]
```diff
@@ -163,15 +459,70 @@
...
+                        );
+                    }
+
+                    // sets the default
+                    $left_select.val(Converter.from_currency);
+                    $right_select.val(Converter.to_currency);
+                }
+
+                /**
+                 * When the user clicks on the field we select
```
Comment (mattr555) [inline share/spice/currency/currency.js]: Not sure how I feel about this functionality. It's pretty awkward when I want to go to insert a character; I have to go to my keyboard and hit an arrow key to do it.
ddgtemp

File: share/spice/currency/currency.js [hunk 2]
```diff
@@ -29,29 +222,150 @@
...
+        getRatesFromAPI: function() {
+
+            var from = $left_select.val();
+            var to = $right_select.val();
+
+            // We'll flip the currencies if the user tries to compare the same symbol
+            if (from === to) {
+                from = Converter.to_currency;
+                to = Converter.from_currency;
+                $right_select.val(to);
```
Comment (mattr555) [inline share/spice/currency/currency.js]: I think you need an analogous line for the left input:
ddgtemp2

File: share/spice/currency/currency.js [hunk 3]
```diff
@@ -29,29 +222,150 @@
...
+        getInverseConversionLabel: function(payload) {
+            return payload["conversion-inverse"]
+        },
+
+        //
+        // Calculates the rates
+        //
+
+        calculateRate: function() {
+            if($currency_input_left.val() !== '' && /\d+/.test($currency_input_left.val())) {
```
Comment (mattr555) [inline share/spice/currency/currency.js]: Minor, but I wonder if we should allow commas in the input. At the moment it just breaks it:
ddgtemp
What if we did a proper format as the user types? i.e. ignore a comma input but automatically insert the commas. (I know this kind of opens up a can of worms about comma vs. decimal formatting in other locales though)

File: share/spice/currency/currency.js [hunk 4]
```diff
@@ -163,15 +459,70 @@
...
+
+                $currency_input_left.keyup(function(_e) {
+                    Converter.calculateRate();
+                });
+
+                $currency_input_right.keyup(function(_e) {
+                    Converter.calculateInverseRate();
+                });
+
+                $selects.change(function(_e) {
```
Comment (mattr555) [inline share/spice/currency/currency.js]: Something a little weird is happening when you do the following:
1) Change left value
2) Change right value
3) Change right currency
4) Go back to previous right currency

ddgtemp3

**Comment by pjhampton** [comment]
Ahh nice. Thanks for the feedback @mattr555. Will address all tomorrow 👍

File: share/spice/currency/currency.js (hunk 3 above)
Comment (pjhampton) [inline share/spice/currency/currency.js]: I have now allowed `,` in the input and break with alphabetical characters like I do conversions IA. It certainly is a can of worms. I have people demanding we support this in the calculator, but to do so we need to support mixed types. I think we need a bigger solution when it comes to treating `,` as `.`

File: share/spice/currency/currency.js (hunk 2 above)
Comment (pjhampton) [inline share/spice/currency/currency.js]: Solved 😄

File: share/spice/currency/currency.js (hunk 1 above)
Comment (pjhampton) [inline share/spice/currency/currency.js]: Similar behavior is exhibited in the [conversions UI](https://duckduckgo.com/?q=unit+converter&atb=v1&ia=answer). I'm trying to create a consistent experience. I'm happy to get into a deeper convo about this though and the greater vision.

File: share/spice/currency/currency.js (hunk 4 above)
Comment (pjhampton) [inline share/spice/currency/currency.js]: This is the expected bavhavour (I think). Correct me if I'm wrong.

Step 1) Start, 5000 usd to xpt = 5.34
Step 2) 50000 usd to xpt = 53.40
Step 3) 50 xpt = 46,806.90 usd
Step 4) Change currency to VND, 46,806.90 usd = 1,043,937.80 vnd

Anything that is updated, the left is considered the **convert from** value (ground truth?). Maybe I'm wrong or just not seeing it.

Comment (mattr555) [inline share/spice/currency/currency.js]: The problem lies in step 5, where the IA now reports that 46,806.90 usd = 0.05 xpt, when it said just before that it's 50 xpt.

File: share/spice/currency/currency.js (hunk 1 above)
Comment (mattr555) [inline share/spice/currency/currency.js]: Ah, ok, but in the conversions UI I can click again once it's highlighted to insert a character. Compare:
ddgtemp
Looking at the code for conversions, I'm not sure what the difference is though 😕.

File: share/spice/currency/currency.js (hunk 4 above)
Comment (pjhampton) [inline share/spice/currency/currency.js]: It's weird, I have gone through all these steps and can't even reproduce that example. Will investigate further in the new week

File: t/Currency.t [hunk 5]
```diff
@@ -455,58 +401,29 @@ ddg_spice_test(
...
-    ),
+    
+    # should no longer work since deprecating the tile view
+    'canada dollar' => undef,
+    '400 euro' => undef,
+    '4k euro' => undef,
+    'bitcoin value' => undef,
+    'bitcoin price' => undef,
+    '100cad' => undef,
+    '100,000cad' => undef,
```
Comment (moollaza) [inline t/Currency.t]: I think these queries should continue to work, defaulting to a comparison with the user's currency:

- https://encrypted.google.com/search?hl=en&q=400%20euro
- https://encrypted.google.com/search?hl=en&q=bitcoin%20price
- https://encrypted.google.com/search?hl=en&q=canada%20dollar

File: share/spice/currency/currency.js (hunk 4 above)
Comment (mattr555) [inline share/spice/currency/currency.js]: Yeah, I'm now unable to reproduce this too. ¯\\\_(ツ)_/¯

File: share/spice/currency/currency.js [hunk 6]
```diff
@@ -1,57 +1,313 @@
...
+            if (from === to) {
+                from = Converter.to_currency;
+                to = Converter.from_currency;
+                $right_select.val(to);
+                $left_select.val(from);
+            }
+
+            var endpoint = "/js/spice/currency/1/" + from + "/" + to;
+            $.get(endpoint, function(payload) {
+
```
Comment (moollaza) [inline share/spice/currency/currency.js]: We might need tpo handle the case of a failure here...
```js
if( !api_result.topConversions.length || api_result.topConversions.length === 0 ) {
    return Spice.failed('currency');
}
```

File: share/spice/currency/currency.js [hunk 7]
```diff
@@ -1,57 +1,313 @@
...
+                // jsonp is returned from the API so we have to alter the contents
+                var response = JSON.parse(payload.trim().replace(/^[^\(]*\(/, '').replace(/\);$/, ''));
```
Comment (moollaza) [inline share/spice/currency/currency.js]: We can use an `alt_to` to prevent this. Seems OK for now.

File: share/spice/currency/currency.js [hunk 8]
```diff
@@ -163,15 +391,54 @@
...
+                    $currency_input_left = $currency.find("#zci--currency-amount-left");
+                    $currency_input_right = $currency.find("#zci--currency-amount-right");
+                    $left_select = $currency.find("#zci--currency-symbol-left");
+                    $right_select = $currency.find("#zci--currency-symbol-right");
+                    $selects = $currency.find("select.zci--currency-symbol");
+                    $more_at_link_normal = $currency.find(".zci__more-at");
+                    $more_at_link_charts = $currency.find(".zci__more-at--info");
+
+                    // apends all the currency names to the selects
+                    for( var i = 0 ; i < currencies.length ; i ++ ) {
```
Comment (moollaza) [inline share/spice/currency/currency.js]: This should probably be part of the template

**Comment by moollaza** [comment]
@pjhampton Thanks! This LGTM :+1:

**Comment by bsstoner** [comment]
This was released to production 🚀

**Pull request merged**